src/fs/driver/devfs/devfs_oldfs.c

```diff
@@ -120,12 +123,114 @@ int devfs_add_char(struct dev_module *cdev) {
 	}
 
 	dev_nas = node.node->nas;
-	dev_nas->fs = filesystem_create("devfs");
-	if (dev_nas->fs == NULL) {
-		return -ENOMEM;
+	dev_nas->fs = devfs_file_system;
+	node.node->nas->fi->privdata = (void *)cdev;
+
+	return 0;
+}
+
+void devfs_fill_inode(struct inode *inode, struct dev_module *devmod, int flags) {
+	assert(inode);
+	assert(devmod);
+
+	inode->nas->fi->privdata = devmod;
+	inode->nas->fs = devfs_file_system;
+	inode->i_mode = flags;
+}
+
+extern struct dev_module **get_cdev_tab();
+/**
+ * @brief Iterate elements of /dev
+ *
+ * @note Devices are iterated type by type
+ * @note Two least significant bits of fs-specific pointer is dev type, the
+ * rest is dev number in dev tab
+ *
+ * @param next Inode to be filled
+ * @param parent Ignored
+ * @param ctx
+ *
+ * @return Negative error code
+ */
+static int devfs_iterate(struct inode *next, struct inode *parent, struct dir_ctx *ctx) {
+	int i;
+	struct block_dev **bdevtab = get_bdev_tab();
+	//struct dev_module *dev_module;
+	struct dev_module **cdevtab = get_cdev_tab();
+
+	i = ((intptr_t) ctx->fs_ctx);
+
+	for (; i < MAX_BDEV_QUANTITY; i++) {
+		if (bdevtab[i]) {
+			ctx->fs_ctx = (void *)(intptr_t)i + 1;
+			devfs_fill_inode(next, bdevtab[i]->dev_module, S_IFBLK | S_IRALL | S_IWALL);
+			//next->length = bdevtab[i]->size;
+			strncpy(next->name, bdevtab[i]->name, sizeof(next->name) - 1);
+			next->name[sizeof(next->name) - 1] = '\0';
+			return 0;
+		}
 	}
 
-	node.node->nas->fi->privdata = (void *)cdev;
+	/* Dynamically allocated devices */
+	for (; i < (MAX_CDEV_QUANTITY + MAX_BDEV_QUANTITY); i++) {
+		if (cdevtab[i - MAX_BDEV_QUANTITY]) {
+			ctx->fs_ctx = (void *) ((intptr_t) i + 1);
+			devfs_fill_inode(next, cdevtab[i - MAX_BDEV_QUANTITY], S_IFCHR | S_IRALL | S_IWALL);
+			strncpy(next->name, cdevtab[i - MAX_BDEV_QUANTITY]->name, sizeof(next->name) - 1);
+			next->name[sizeof(next->name) - 1] = '\0';
+			return 0;
+		}
+	}
```

Comment:
May be use `for (i = 0; i < MAX_CDEV_QUANTITY; i++) {` to get rid of `cdevtab[i - MAX_BDEV_QUANTITY]` and just use `cdevtab[i]`?

Reply:
Ah, now I see we have common counter for both bdev and cdev.